### PR TITLE
Add HelloWorld sample of LLVMSharp15

### PR DESCRIPTION
**Pull request by CWKSC**
This HelloWorld sample will output:

```ll
[Output of LLVM IR]:

; ModuleID = 'Hello World'
source_filename = "Hello World"

@0 = private unnamed_addr constant [13 x i8] c"Hello World!\00", align 1

declare i32 @puts(ptr %0)

define i32 @main() {
entry:
    %0 = call i32 @puts(ptr @0)
    ret i32 0
}

[Output of JIT Compiler]:

Hello World!
```

### CODE REVIEW

**Comment by tannergooding** [comment]
Sorry for the delay in getting this reviewed, was working through some infrastructure issues and upgrading to a new code signing client.

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright © Tanner Gooding and Contributors. Licensed under the MIT License (MIT). See License.md in the repository root for more information.
+
+using System.Text;
+using LLVMSharp.Interop;
+
+namespace HelloWorld;
+
+public static class LLVMSharpUtil
+{
+
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:10]: nit: Stray newline

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    // Constant //
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:11]: nit: These comments shouldn't have a trailing `//`

Several of them seem like they are "obvious" and clearly covered by the member name. For the ones that aren't, they'd be better represented as `documentation comments` -- e.g. `/// <summary>The integer constant 0.</summary>`

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    public static readonly unsafe LLVMOpaqueValue* ConstInt0 = LLVM.ConstInt(LLVM.Int32Type(), 0, 1);
+
+    // Convert string to sbyte* //
+    public static unsafe sbyte* ToSbytePointer(this string value)
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:15]: This method is not safe and represents a GC hole.

Same goes for basically any of these functions that `fix` a managed type and then return the pointer.

There are some APIs that do this correctly here: https://github.com/dotnet/LLVMSharp/tree/main/sources/LLVMSharp.Interop/Internals

You would use it as:
```csharp
using var marshaledString= new MarshaledString("Text");
ApiThatNeedsAStringPointer(marshaledString);
```

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    // Sometimes crashes with "LLVM ERROR: IMAGE_REL_AMD64_ADDR32NB relocation requires an ordered section layout"
+    public static unsafe void RunMain(LLVMOpaqueModule* module)
+    {
+        LLVM.LinkInMCJIT();
+        LLVM.InitializeX86TargetMC();
+
+        LLVM.InitializeX86Target();
+        LLVM.InitializeX86TargetInfo();
+        LLVM.InitializeX86AsmParser();
+        LLVM.InitializeX86AsmPrinter();
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:137]: These should probably be `LLVM.IniitalizeNativeTarget()` and `LLVM.InitializeNativeAsmParser/Printer`

or alternatively `LLVM.InitializeAllTargets` and `LLVM.InitializeAllAsmParers/Printers`

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        CreateExecutionEngineForModule(
+            out var engine, out _,
+            module
+        );
+        var main = LLVM.GetNamedFunction(module, "main".ToSbytePointer());
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:143]: This be `"main"u8` which would be more efficient than taking a UTF16 string, converting to UTF8, and then passing it along.

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        _ = LLVM.RunFunction(engine, main, 0, null);
+    }
+
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:146]: nit: Stray newline:
```suggestion
```

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        out sbyte* error,
+        LLVMOpaqueModule* module
+    )
+    {
+        LLVMOpaqueExecutionEngine* outEE = null;
+        sbyte* outError = null;
+        _ = LLVM.CreateExecutionEngineForModule(&outEE, module, &outError);
+        engine = outEE;
+        error = outError;
+    }
```
Comment (tannergooding) [inline samples/HelloWorld/LLVMSharpUtil.cs:125]: A lot of this is unnecessary and already exists via the "type safe extensions".

For example, this is exposed on `LLVMModuleRef` as `LLVMExecutionEngineRef CreateExecutionEngine()` or as `bool TryCreateExecutionEngine(out LLVMExecutionEngineRef OutEE, out string OutError)`

File: samples/HelloWorld/Program.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    // Declare puts() and main()
+    LLVMSharpUtil.DeclarePuts(
+        out var putsFunc, out var putsFuncTy,
+        module
+    );
+    LLVMSharpUtil.CreateMain(
+        out var _, out var mainBuilder,
+        module
+    );
```
Comment (tannergooding) [inline samples/HelloWorld/Program.cs:21]: These ones aren't really utilities and should probably exist in `Program.cs`

File: samples/HelloWorld/Program.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+using HelloWorld;
+using LLVMSharp.Interop;
+
+unsafe
+{
+
+    // Context and Module //
+    var context = LLVM.ContextCreate();
+    var module = LLVM.ModuleCreateWithNameInContext("Hello World".ToSbytePointer(), context);
```
Comment (tannergooding) [inline samples/HelloWorld/Program.cs:11]: If you start the program using this instead, then you get the type safe extensions which make this a lot nicer/easier:
```csharp
var context = LLVMContextRef.Create();
var module = context.CreateModuleWithName("Hello World");
```

File: samples/HelloWorld/Program.cs (hunk 8 above)
Comment (tannergooding) [inline samples/HelloWorld/Program.cs:21]: These ones aren't really utilities and should probably exist in Program.cs

File: samples/HelloWorld/LLVMSharpUtil.cs (hunk 3 above)
Comment (tannergooding) [reply samples/HelloWorld/LLVMSharpUtil.cs:15]: That being said, needing such wrappers here should be unnecessary as there exists a "type safe" way of doing all this natively. I've given some examples further down.

File: samples/HelloWorld/LLVMSharpUtil.cs (hunk 7 above)
Comment (perlun) [reply samples/HelloWorld/LLVMSharpUtil.cs:125]: Sounds good, thanks for this suggestion @tannergooding! Working my way through getting the Kaleidoscope example working on recent LLVMSharp versions, and both this MR and that comment is a useful help in this.

As a slight side note, doesn't this code (in latest `main`) leak the unmanaged `pError` memory?

https://github.com/dotnet/LLVMSharp/blob/b10edc6ed2c10479f62313a4c4baf82747a36bb6/sources/LLVMSharp.Interop/Extensions/LLVMModuleRef.cs#L323-L340

Shouldn't this be freed using `LLVM.DisposeMessage` before this method returns, or am I misreading things here? :thinking:

File: samples/HelloWorld/LLVMSharpUtil.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+    // Convert LLVMOpaqueValue*[] to LLVMOpaqueValue** //
+    public static unsafe LLVMOpaqueValue** CreateLLVMOpaqueValueList(params LLVMOpaqueValue*[] values)
+    {
+        fixed (LLVMOpaqueValue** temp = values)
+        {
+            return temp;
+        }
+    }
```
Comment (perlun) [inline samples/HelloWorld/LLVMSharpUtil.cs:41]: @tannergooding - any "type safe" way to do this (create a `LLVMOpaqueValue**`) with the wrappers you mention above? :thinking: Still struggling to get a basic example working with LLVMSharp15, and this is one of the remaining hurdles.

**Pull request closed without merging**